File: backend/app/cnpj.py
"""Normalização e validação de CNPJ.

O banco guarda o CNPJ formatado (``VARCHAR(18)`` — ``00.000.000/0000-00``),
então toda entrada passa por aqui antes de chegar ao modelo. Comparações
sempre usam :func:`cnpj_digitos`, nunca a string formatada.
"""

CNPJ_TAMANHO = 14
CNPJ_FORMATADO_TAMANHO = 18


def cnpj_digitos(valor: str | None) -> str:
    """Devolve apenas os dígitos do CNPJ (string vazia quando não há valor)."""
    if not valor:
        return ""
    return "".join(c for c in valor if c.isdigit())


def cnpj_valido(valor: str | None) -> bool:
    """CNPJ tem 14 dígitos e não pode ser uma repetição do mesmo dígito."""
    digitos = cnpj_digitos(valor)
    if len(digitos) != CNPJ_TAMANHO:
        return False
    return len(set(digitos)) > 1


def formatar_cnpj(valor: str | None) -> str:
    """Formata como ``00.000.000/0000-00``; devolve o original se não der."""
    d = cnpj_digitos(valor)
    if len(d) != CNPJ_TAMANHO:
        return (valor or "").strip()
    return f"{d[:2]}.{d[2:5]}.{d[5:8]}/{d[8:12]}-{d[12:]}"

File: frontend/src/lib/cnpj.ts
/** Espelha backend/app/cnpj.py — o banco guarda o CNPJ formatado (VARCHAR(18)). */

export const CNPJ_DIGITOS = 14;
export const CNPJ_TAMANHO_FORMATADO = 18;

export function apenasDigitos(valor: string): string {
  return valor.replace(/\D/g, "");
}

/** Aplica a máscara 00.000.000/0000-00 e descarta dígitos excedentes. */
export function mascararCnpj(valor: string): string {
  const d = apenasDigitos(valor).slice(0, CNPJ_DIGITOS);
  if (d.length <= 2) return d;
  if (d.length <= 5) return `${d.slice(0, 2)}.${d.slice(2)}`;
  if (d.length <= 8) return `${d.slice(0, 2)}.${d.slice(2, 5)}.${d.slice(5)}`;
  if (d.length <= 12) return `${d.slice(0, 2)}.${d.slice(2, 5)}.${d.slice(5, 8)}/${d.slice(8)}`;
  return `${d.slice(0, 2)}.${d.slice(2, 5)}.${d.slice(5, 8)}/${d.slice(8, 12)}-${d.slice(12)}`;
}

export function cnpjValido(valor: string): boolean {
  const d = apenasDigitos(valor);
  // 14 dígitos e não pode ser o mesmo dígito repetido (00.000.000/0000-00).
  return d.length === CNPJ_DIGITOS && new Set(d).size > 1;
}
